### ISSUE TITLE
order of the calculateCoefficients reults

### ISSUE THREAD

**Issue by lalawow** [omphalos/smr#2]
Hi, I am trying the smr v0.10.0

I used next codes to test the smr calculateCoefficients function

> var smr = require('smr')
> var regression1 = new smr.Regression({ numX: 3, numY: 1 })
> 
> regression1.push({ x: [1/3,2,0.5], y: [2] })
> regression1.push({ x: [2/3,4,0.5], y: [3] })
> regression1.push({ x: [3/3,7,0.5], y: [4] })
> 
> console.log("test1",regression1.calculateCoefficients()) //return [[3],[0],[2]] instead of [[0],[2],[3]]
> 
> var regression2 = new smr.Regression({ numX: 2, numY: 1 })
> 
> regression2.push({ x: [1/3,0.5], y: [2] })
> regression2.push({ x: [2/3,0.5], y: [3] })
> regression2.push({ x: [3/3,0.5], y: [4] })
> console.log("test2",regression2.calculateCoefficients()) //return [[3],[2]]
> 
> var regression3 = new smr.Regression({ numX: 4, numY: 1 })
> 
> regression3.push({ x: [1/3,2,0.5,8], y: [2] })
> regression3.push({ x: [2/3,4,0.5,3], y: [3] })
> regression3.push({ x: [3/3,7,0.5,2], y: [4] })
> 
> console.log("test3",regression3.calculateCoefficients()) //return [[3],[0],[2],[0]] instead of others

when NumX=2, the result is ok (test2)
when NumX=3, the order of the result is weird(test1)
when NumX=4, the result dosen't make sense

my test space is using nodejs v4.1.1

**Comment by omphalos** [comment]
Hi @lalawow, thanks for opening this issue.  Sorry it's taken me a few days to get to this.

In your examples, the results look okay to me.  Here is a function you can use to test `smr`'s output when you have `numY=1`:

``` js
function test(options) {
  var coefficients = options.coefficients
  var x = options.x
  var y = options.y
  var sum = 0
  coefficients.forEach((coefficient, i) => sum += coefficient[0] * x[i])
  console.log('predicted=' + sum + ' y=' + y)
}
```

Running this on your examples:

``` js
test({ coefficients: [[3],[0],[2]], x: [1/3,2,0.5], y: [2] }) // outputs predicted=2 y=2
test({ coefficients: [[3],[0],[2]], x: [2/3,4,0.5], y: [3] }) // outputs predicted=3 y=3
test({ coefficients: [[3],[0],[2]], x: [3/3,7,0.5], y: [4] }) // outputs predicted=4 y=4

test({ coefficients: [[3],[2]], x: [1/3,0.5], y: [2] }) // outputs predicted=2 y=2
test({ coefficients: [[3],[2]], x: [2/3,0.5], y: [3] }) // outputs predicted=3 y=3
test({ coefficients: [[3],[2]], x: [3/3,0.5], y: [4] }) // outputs predicted=4 y=4

test({ coefficients: [[3],[0],[2],[0]], x: [1/3,2,0.5,8], y: [2] }) // outputs predicted=2 y=2
test({ coefficients: [[3],[0],[2],[0]], x: [2/3,4,0.5,3], y: [3] }) // outputs predicted=3 y=3
test({ coefficients: [[3],[0],[2],[0]], x: [3/3,7,0.5,2], y: [4] }) // outputs predicted=4 y=4
```

In all these cases the results look good to me.  The predicted output exactly matches the y values.

I hope that helps.  I may have misunderstood your question.  Let me know if I'm missing something.

**Comment by omphalos** [comment]
@lalawow I see what you're saying - smr isn't returning the right coefficients.  I'll look into this some more.

**Comment by lalawow** [comment]
hi @omphalos 
After the smr.js issue, I checked numeric.js, and tried to apply the Normal Equation of the multiple regression by numeric.js directly. It's fine. Hope this can help u.

my simple test code:

```
var numeric = require("numeric")
var x=[[1/3,2,0.5],[2/3,4,0.5],[3/3,7,0.5],[4/3,5,0.5]]
var y=[2,3,4,5]
var res = numeric.dot(numeric.dot(numeric.inv(numeric.dot(numeric.transpose(x),x)),numeric.transpose(x)),y)
console.log(res)
```

**Comment by omphalos** [comment]
@lalawow Thanks, I'm not sure if that works for the third example though.  This is the output I get:

``` js
var x = [[1/3,2,0.5,8],[2/3,4,0.5,3],[3/3,7,0.5,2]]
var y = [2,3,4]
numeric.dot(numeric.dot(numeric.inv(numeric.dot(numeric.transpose(x),x)),numeric.transpose(x)),y)
// output:
// [-3.5, 0.75, 8.5, -0.125]
```

I found a good implementation which I cloned [here](https://github.com/omphalos/rosettacode-js-regression).  It's going to take a little work for me to make that incremental though.

**Comment by lalawow** [comment]
@omphalos 
I see. When we use Normal Equation to solve multiple regression, we need the length of Xarray is >= numX. Other wise, the solve of the multiple regression is not supposed to be a only one, and the normal equation will give us weird answers.

so for the third exmple I gave before, I should insert another element into X and Y. like [4/3,7,0.5,7],[5]
I got the smr output: [0,0,3,2]
It's similar to example 1: the numbers are right, but in wrong order.

**Comment by omphalos** [comment]
Thanks @lalawow.  I switched to `numeric.inv` and added tests with your examples, which now return the expected values.  Would you mind verifying?

**Comment by lalawow** [comment]
I am happy to verify it.

**Comment by lalawow** [comment]
I think the issue was solved by this v0.12.3
I used one more example to test it. It's fine.

```
var regression4 = new smr.Regression({ numX: 4, numY: 1 })

regression4.push({ x: [1/3,1/4,1/5,8], y: [3] })
regression4.push({ x: [2/3,3/4,1/5,3], y: [6] })
regression4.push({ x: [3/3,4/4,1/5,2], y: [8] })
regression4.push({ x: [4/3,2/4,1/5,7], y: [7] })

console.log("test4",regression4.calculateCoefficients()) //return [[3],[4],[5],[0]] 
```

**Comment by omphalos** [comment]
Thanks for your help @lalawow.  I'm going to close this issue.
